[Low Xi Si] iP

LowXiSi opened the pull request:
# LoafyPro

LoafyPro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- fast to use

All you need to do is,

1. download it from [here](https://github.com/LowXiSi/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you.

And it is **FREE**! 🤩 🤩 🤩 

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
    public static void main(String[] args) {
        new Loafy("./data/loafy.txt").run();
    }
```

Here is what users have to say about LoafyPro
> Frees up so much space in my brain that my CS2103 grade went up a rank!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

LowXiSi committed:
Replace Duke with Loafy

* Rename main class to Loafy
* Add greeting and exit statements

LowXiSi committed:
Add support for echo

* Take in user input
* Update chat display format
* Echo user command
* End chat on "bye" command

LowXiSi committed:
Add Task class

LowXiSi committed:
Add TaskList class

* add(Task task) method to add tasks
* view() method to view current list

LowXiSi committed:
Support for add task and view list

*Store user input into a list
*Recognise command "list" to output current list

LowXiSi committed:
Add setter for task status

LowXiSi committed:
Add markTask method in TaskList class

LowXiSi committed:
Support for task marking

LowXiSi committed:
Support for Todo, Deadline, Event

LowXiSi committed:
Make Task class abstract

LowXiSi committed:
Clean up Loafy class

LowXiSi committed:
Implement automated text UI testing

LowXiSi committed:
Add LoafyException class

*contains all error messages for possible errornous user input

LowXiSi committed:
Handle user input errors

LowXiSi committed:
Support for delete

LowXiSi committed:
Support for saving data to hard drive

LowXiSi committed:
Ignore data folder

LowXiSi committed:
Support for loading data from previous run

LowXiSi committed:
Merge branch 'branch-Level-7'

LowXiSi committed:
Deadline and Event to store LocalDateTime dates

LowXiSi committed:
Merge branch 'branch-Level-8'

LowXiSi committed:
Extract functionalities into classes

LowXiSi committed:
Merge branch 'branch-A-MoreOOP'

LowXiSi committed:
Fix minor bugs and error messages

LowXiSi committed:
Clean up coding standard

LowXiSi committed:
Sort classes into packages

LowXiSi committed:
Merge branch 'add-gradle-support'

LowXiSi committed:
Use Gradle

LowXiSi committed:
Add JUnit tests for Parser and Deadline

LowXiSi committed:
Add JavaDoc comments

LowXiSi committed:
Tweak code to comply with coding standard

LowXiSi committed:
Add support for find

LowXiSi committed:
Merge branch 'branch-A-JavaDoc'

LowXiSi committed:
Merge branch 'branch-A-CodingStandard'

LowXiSi committed:
Merge branch 'branch-Level-9'

YINZE66 commented on src/main/java/Loafy/parser/Parser.java:
```diff
@@ -0,0 +1,115 @@
+
+import java.util.Arrays;
+
+import loafy.command.AddCommand;
+import loafy.command.Command;
+import loafy.command.DeleteCommand;
+import loafy.command.ExitCommand;
+import loafy.command.FindCommand;
+import loafy.command.ListCommand;
+import loafy.command.MarkCommand;
```
I like how you import the specific classes instead of wildcard 👍

YINZE66 commented on src/main/java/Loafy/parser/Parser.java:
```diff
@@ -0,0 +1,115 @@
+
+
+public class Parser {
+    public static Command parse(String line) throws LoafyException {
```
Maybe should add JavaDoc for the parse method since it's quite complicated? 👀

YINZE66 commented on src/main/java/Loafy/task/Event.java:
```diff
@@ -0,0 +1,69 @@
+        this.endDate = endDate;
+    }
+
+    /**
+     * Returns a string representation of this event task for user view.
+     * The string representation is in the format "[E][status] name (from: start date to: end date)",
+     * "[E]" symbolises type of task (event);
+     * status is represented by "X" if task is done and " " if task is undone;
+     * start and end dates are in "d/M/yyyy HHmm" format.
+     *
+     * @return a string representation of this event task.
+     */
```
Like the method description here, very comprehensible! 🚀

YINZE66 commented on src/main/java/Loafy/ui/Ui.java:
```diff
@@ -0,0 +1,28 @@
+        this.reply("Hmm...\n" +
+                exception.getMessage() +
+                "\n   Try again!");
+    }
```
You are using indentation correctly, nice job! 💯

YINZE66 commented on src/main/java/Loafy/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+
+import loafy.tasklist.TaskList;
+import loafy.ui.Ui;
+
```
Maybe add class description for the abstract command class since it has many subclasses? ❓

YINZE66 reviewed (commented):
Overall, your code follow the coding standard well! Almost no violation I can find. Nice job!! 👍

DINESH1201 commented on src/main/java/Loafy/command/FindCommand.java:
```diff
@@ -0,0 +1,17 @@
+import loafy.tasklist.TaskList;
+import loafy.ui.Ui;
+
+public class FindCommand extends Command {
```
I think you might have missed out this file while doing the JavaDoc. Just to let you know ;)

DINESH1201 commented on src/main/java/Loafy/parser/Parser.java:
```diff
@@ -0,0 +1,115 @@
+            int i = Arrays.asList(arr).indexOf("/by");
+            if (i == -1) {
+                throw LoafyException.ofNoDeadline();
+            } else {
+                String name = joinRange(arr, 1, i);
+                String date = joinRange(arr, i + 1, arr.length);
+                if (name.isEmpty() || date.isEmpty()) {
+                    throw LoafyException.ofNoDeadline();
+                } else {
+                    LocalDateTime dateTime = parseDate(date);
+                    Task task = new Deadline(name, dateTime);
+                    return new AddCommand(task);
+                }
+            }
+        } else if (arr[0].equals("event")) {
+            int fromIndex = Arrays.asList(arr).indexOf("/from");
+            int toIndex = Arrays.asList(arr).indexOf("/to");
+            if (fromIndex == -1 || toIndex == -1) {
+                throw LoafyException.ofNoEventDates();
+            } else {
+                String name = joinRange(arr, 1, fromIndex);
+                String startDateString = joinRange(arr, fromIndex + 1, toIndex);
+                String endDateString = joinRange(arr, toIndex + 1, arr.length);
+                if (name.isEmpty() || startDateString.isEmpty() || endDateString.isEmpty()) {
+                    throw LoafyException.ofNoEventDates();
+                } else {
+                    LocalDateTime startDate = parseDate(startDateString);
+                    LocalDateTime endDate = parseDate(endDateString);
+                    Task task = new Event(name, startDate, endDate);
+                    return new AddCommand(task);
+                }
+            }
+        } else if (arr[0].equals("find")) {
+            if (arr.length == 1) {
+                throw LoafyException.ofEmptyFind();
+            }
+            return new FindCommand(arr[1]);
+        } else {
+            throw LoafyException.ofInvalidCommand();
+        }
```
Any reason why you did it using if else statements instead of a switch case?

DINESH1201 commented on src/main/java/Loafy/storage/Storage.java:
```diff
@@ -0,0 +1,83 @@
+import loafy.task.Task;
+import loafy.task.Todo;
+
+public class Storage {
```
Maybe you could include some JavaDoc for the Storage class

DINESH1201 reviewed (commented):
I like that your code is quite comprehensive overall. Your formatting is well done!!

LowXiSi committed:
Use checkStyle to detect coding style violations

LowXiSi committed:
Fix minor bug

LowXiSi committed:
Add GUI for Loafy

LowXiSi committed:
Merge branch 'branch-Level-10'

LowXiSi committed:
Parser class: Extract command handling out from parse method

The parse method in the Parser class is very long and unreadable.

Extracting command handling improves the code quality.

Let's,
* extract command handling into individual methods
* add JavaDoc for each of the methods

LowXiSi committed:
Command subclasses: Update JavaDoc messages

JavaDoc messages do not match changed methods.

This is confusing and makes code less readable.

Let's update JavaDoc to match the edited methods.

LowXiSi committed:
Add assertions to check important assumptions

LowXiSi committed:
Storage class: Add JavaDoc and highlight "happy path"

LowXiSi committed:
Merge pull request #2 from LowXiSi/branch-A-Assertions

Add assertions

LowXiSi committed:
Merge branch 'branch-A-CodeQuality'

LowXiSi committed:
Support more natural date formats

LowXiSi committed:
Merge branch 'branch-BCD-Extension'

LowXiSi committed:
Event command: Enforce precedence of "/from" before "/to"

LowXiSi committed:
Add support for program termination

LowXiSi committed:
Update readme and images

LowXiSi committed:
Update readme to contain heading links

LowXiSi committed:
Reconfigure the links for readme

LowXiSi committed:
Edit readme format to enable links

LowXiSi committed:
Fix code style

LowXiSi committed:
Update version number

LowXiSi committed:
Update invalid command error message

LowXiSi committed:
Update invalid command error message to make list clearer

LowXiSi committed:
Update readme to contain mark and unmark
